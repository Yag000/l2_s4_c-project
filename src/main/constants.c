#include "constants.h"

bool interactive = false;

const int EXIT_PROGRAM_SUCCESS = -2;  // Error code that indicates that the program executed successfully and that
                                      // it should be terminated
const int SUCCESS = 0;                // Error code that indicates that the program executed successfully
const int FATAL_ERROR = -1;           // Error code that indicates that he program should stop
const int INVALID_NAME = 1;           // Error code that indicates that the name of a node is invalid
const int INVALID_PATH = 2;           // Error code that indicates that the path of a node is invalid
const int INVALID_NUMBER_OF_ARGS = 3; // Error code that indicates that the number of arguments is invalid
const int INVALID_SELECTION = 4;      // Error code that indicates that the element selected by the user is invalid
