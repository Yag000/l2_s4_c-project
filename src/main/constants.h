#ifndef CONSTANTS_H
#define CONSTANTS_H

#include <stdio.h>
#include <stdbool.h>

extern FILE *out_stream;
extern bool verbose;

#endif