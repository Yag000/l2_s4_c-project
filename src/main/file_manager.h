#ifndef FILE_MANAGER_H
#define FILE_MANAGER_H

#include <stdio.h>

FILE *open_file(const char *path, const char *mode);
int close_file(FILE *file, const char *path);
int reset_file(const char *path);

#endif