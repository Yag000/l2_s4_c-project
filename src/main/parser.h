#ifndef PARSER_H
#define PARSER_H


int parse_file(char *path);
int parse_line(char *line);

#endif