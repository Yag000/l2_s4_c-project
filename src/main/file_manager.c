#include <stdlib.h>
#include <stdio.h>
#include <assert.h>
#include <string.h>

#include "file_manager.h"
#include "constants.h"

static void print_error_message(const char *, const char *);

/*
Opens the stream, send an error message if it fails
*/
FILE *open_file(const char *path, const char *mode)
{
    FILE *file = fopen(path, mode);
    if (file == NULL)
    {
        print_error_message("Problème ouverture fichier ", path);
    }
    return file;
}

/*
Closes the stream, send an error message if it fails
*/
int close_file(FILE *file, const char *path)
{
    int exit_code = fclose(file);

    if (exit_code != 0)
    {
        print_error_message("Problème fermeture fichier ", path);
        return FATAL_ERROR;
    }
    return SUCCESS;
}

/*
Prints the error message and the path of the file where it happened an error
*/
static void print_error_message(const char *error_message, const char *path)
{
    char *message = malloc(sizeof(char) * (strlen(error_message) + strlen(path) + 1));
    assert(message != NULL);
    message = strcat(message, error_message);
    message = strcat(message, path);
    perror(message);
    free(message);
}
