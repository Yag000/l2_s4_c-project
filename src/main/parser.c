#define _GNU_SOURCE
#include <assert.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "command.h"
#include "constants.h"
#include "file_manager.h"
#include "parser.h"
#include "string_utils.h"

/*
The maximum number of arguments for a command.
We could also use a dynamic array to store the arguments,
but for the purpose of this project, we will use a fixed size array of 2.
*/
#define MAX_COMMAND_ARGUMENTS 2

static command *get_command_from_iterator(string_iterator *);

/*
Parses a file containing commands.
The file parses lines up to MAX_LINE_LENGTH characters.
It returns 0 if the execution of the commands is successful.
*/
int parse_file(const char *path)
{
    FILE *file = open_file(path, "r");

    if (file == NULL)
    {
        return -1;
    }

    int exit_code = 0;

    if (verbose)
    {
        fputs("Parsing file ", out_stream);
        fputs(path, out_stream);
        fputs(" ...\n", out_stream);
    }

    char *line = NULL;
    size_t len = 0;
    ssize_t read;
    while ((read = getline(&line, &len, file)) != -1)
    {
        exit_code = parse_line(line);
        if (exit_code != 0)
        {
            perror("Probleme parse line");
            break;
        }
    }

    free(line);
    close_file(file, path);

    return exit_code;
}

/*
Parses a line of the file.
It returns 0 if the command execution is successful.
*/
int parse_line(char *line)
{
    string_iterator *iterator = create_string_iterator(line, ' ');

    if (iterator == NULL)
    {
        perror("Problème initialisation iterator");
        free(line);
        return -1;
    }

    command *command = get_command_from_iterator(iterator);

    if (command == NULL)
    {
        perror("Problème creation commande");
        free(line);
        return -1;
    }

    int exit_code = execute_command(command);

    destroy_string_iterator(iterator);
    destroy_command(command);

    return exit_code;
}

/*
Returns a command from a string iterator.
*/
static command *get_command_from_iterator(string_iterator *iterator)
{
    if (!has_next_word(iterator))
    {
        perror("Problème iterator vide");
        return NULL;
    }

    char *command_ = next_word(iterator);

    char **args = malloc(sizeof(char *) * MAX_COMMAND_ARGUMENTS);
    assert(args != NULL);

    int args_number = 0;
    while (has_next_word(iterator) && args_number < MAX_COMMAND_ARGUMENTS)
    {
        args[args_number] = strip_newline(next_word(iterator));
        args_number++;
    }

    if (args_number < MAX_COMMAND_ARGUMENTS)
    {
        if (args_number == 0)
        {
            free(args);
            args = NULL;
        }
        else
        {
            args = realloc(args, sizeof(char *) * args_number);
        }
    }

    return create_command(command_, args_number, args);
}
