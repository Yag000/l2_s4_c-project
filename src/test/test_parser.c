#include <stdio.h>

#include "../main/constants.h"
#include "../main/file_manager.h"
#include "../main/parser.h"
#include "test_core.h"

static void test_parse_file(test_info *);

test_info *test_parser()
{
    // Test setup
    print_test_header("parser");
    clock_t before = clock();
    test_info *info = create_test_info();

    // Add tests here
    test_parse_file(info);

    // End of tests
    info->time = clock_ticks_to_seconds(clock() - before);
    print_test_footer("parser", info);
    return info;
}

static void test_parse_file(test_info *info)
{
    out_stream_path = "src/resources/unit_tests/output/test_parser.txt";
    int error_code = -1;

    out_stream = open_file(out_stream_path, "w");

    if (out_stream == NULL)
    {
        return;
    }

    error_code = parse_file("src/resources/unit_tests/input/test_parser.txt");

    close_file(out_stream, out_stream_path);

    out_stream = stdout;

    handle_boolean_test(error_code == 0, true, __LINE__, __FILE__, info);
}
