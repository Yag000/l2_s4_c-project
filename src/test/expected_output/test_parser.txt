Parsing file src/test/input/test_parser.txt ...
$ debug

$ debug one
one
$ debug one two
one two
$ debug one two
one two
$ debug one two
one two
$ debug one two
one two
$ debug one two
one two
